Allow marking arguments as constant in the function signature

duanmeng opened the pull request:
Details in #4140

mbasmanova commented on velox/expression/FunctionSignature.h:
```diff
@@ -104,9 +104,18 @@ class TypeSignature {
+    return isConstant_;
+  }
+
+  void setIsConstant(bool isConstant) {
```
It would be nice to keep this class immutable. Since is-constant is not a property of the type, consider, adding a list of boolean is-constant flags to the FunctionSignature class.

```
  const TypeSignature returnType_;
  const std::vector<TypeSignature> argumentTypes_;
  const std::vector<bool> constantArguments_;
  const bool variableArity_;
```

duanmeng replied:
Fixed

mbasmanova commented on velox/expression/tests/ArgumentTypeFuzzerTest.cpp:
```diff
@@ -73,22 +73,31 @@ TEST_F(ArgumentTypeFuzzerTest, concreteSignature) {
                          .build();
 
-    testFuzzingSuccess(signature, BIGINT(), {VARCHAR(), ARRAY(BOOLEAN())});
+    testFuzzingSuccess(
```
The change here should not affect ArgumentTypeFuzzer, hence, I'm surprised to see changes to ArgumentTypeFuzzerTest.cpp. Let's remove these. Let's add tests to velox/exec/tests/FunctionSignatureBuilderTest.cpp instead.

duanmeng replied:
Fixed

mbasmanova commented:
@duanmeng Should we also update FunctionSignature::toString() to include "const" qualified for the arguments?

duanmeng commented:
> @duanmeng Should we also update FunctionSignature::toString() to include "const" qualified for the arguments?

@mbasmanova This may be confused with the *const* in C++, or should we use *constant* as close to the semantics of velox constant vector?

mbasmanova commented:
> This may be confused with the _const_ in C++, or should we use _constant_ as close to the semantics of velox constant vector?

@duanmeng I don't have a strong preference here. Here are some options.

```
(bigint, constant double) -> double

(bigint, const double) -> double

(bigint, fixed double) -> double
```

duanmeng commented:
> > This may be confused with the _const_ in C++, or should we use _constant_ as close to the semantics of velox constant vector?
> 
> @duanmeng I don't have a strong preference here. Here are some options.
> 
> ```
> (bigint, constant double) -> double
> 
> (bigint, const double) -> double
> 
> (bigint, fixed double) -> double
> ```

@mbasmanova I've resolved all the comments and updated the toString method, could you help to take a look?

mbasmanova commented on velox/expression/FunctionSignature.cpp:
```diff
@@ -199,6 +211,11 @@ void validate(
+  VELOX_USER_CHECK_EQ(
+      argumentTypes.size(),
+      constantArguments.size(),
+      "Argument types size is not equal to constant flags");
```
Nice check.

mbasmanova commented on velox/exec/tests/FunctionSignatureBuilderTest.cpp:
```diff
@@ -177,3 +177,30 @@ TEST_F(FunctionSignatureBuilderTest, anyInReturn) {
       "Type 'Any' cannot appear in return type");
 }
+
+TEST_F(FunctionSignatureBuilderTest, constFalgs) {
```
typos: constFalgs -> constantFlags

mbasmanova commented on velox/exec/tests/FunctionSignatureBuilderTest.cpp:
```diff
@@ -177,3 +177,30 @@ TEST_F(FunctionSignatureBuilderTest, anyInReturn) {
+  EXPECT_TRUE(signature->constantArguments().at(1));
+  EXPECT_FALSE(signature->constantArguments().at(2));
+  EXPECT_EQ(
+      "(double,constant boolean,bigint) -> bigint", signature->toString());
```
Nice test. Would you also add tests for constant variadic argument and constant generic argument?

mbasmanova commented on velox/exec/tests/FunctionSignatureBuilderTest.cpp:
```diff
@@ -177,3 +177,30 @@ TEST_F(FunctionSignatureBuilderTest, anyInReturn) {
+  EXPECT_EQ(
+      "(double,constant boolean,bigint) -> bigint", signature->toString());
+
+  auto aggSignature = AggregateFunctionSignatureBuilder()
```
Consider moving this into a separate test and renaming to 'signature'.

```
TEST_F(FunctionSignatureBuilderTest, scalarConstantFlags) {

TEST_F(FunctionSignatureBuilderTest, aggregateConstantFlags) {
```

mbasmanova reviewed (commented):
@duanmeng Looks great. Let's add tests for variadic and generic arguments and then merge.

mbasmanova commented on velox/exec/tests/FunctionSignatureBuilderTest.cpp:
```diff
@@ -177,3 +177,74 @@ TEST_F(FunctionSignatureBuilderTest, anyInReturn) {
+    EXPECT_FALSE(signature->constantArguments().at(0));
+    EXPECT_TRUE(signature->constantArguments().at(1));
+    EXPECT_FALSE(signature->constantArguments().at(2));
+    std::cout << signature->toString() << std::endl;
```
Accidental?

duanmeng replied:
Removed, sorry not to check and remove it before the push.

laithsakka commented:
If an argument is marked const in the signature, is the expression compiler allowed to compile such a function if the input is not literal? I do not see changes related to that ?

could you also add a summary explaining the functionalities.

mbasmanova commented:
> If an argument is marked const in the signature, is the expression compiler allowed to compile such a function if the input is not literal? I do not see changes related to that ?
> 
> could you also add a summary explaining the functionalities.

@laithsakka It would nice to add a check for 'const' arguments to the ExprCompiler.  Let's do that in a follow-up PR though to keep PRs small and easy to review.
